types/src/transaction/webauthn.rs

```
@@ -0,0 +1,681 @@
+    ) -> Self {
+        PartialAuthenticatorAssertionResponse {
+            signature: webauthn_signature,
+            authenticator_data,
+            client_data_json,
+        }
+    }
+
+    /// Uses BCS serialization
+    pub fn to_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&self).expect("Only unhandleable errors happen here.")
+    }
+
+    pub fn signature(&self) -> &WebAuthnSignature {
+        &self.signature
+    }
+
+    pub fn signature_bytes(&self) -> Vec<u8> {
+        bcs::to_bytes(&self.signature).expect("Only unhandleable errors happen here.")
+    }
+
+    /// In our adaptation of WebAuthn, the `challenge` provided to `authenticatorGetAssertion`
+    /// is the SHA3-256 digest of the `RawTransaction`.
+    ///
+    /// This function should do the following:
+    /// 1. Verify `actual_challenge` and expected challenge from message are equal
+    /// 2. Construct `verification_data` as the binary concatenation of
+    ///    authenticator_data and SHA-256(client_data_json)
+    /// 3. Signature verification
+    ///
+    /// See WebAuthn §6.3.3 `authenticatorGetAssertion` for more info
+    pub fn verify<T: Serialize + CryptoHash>(
+        &self,
+        message: &T,
+        public_key: &AnyPublicKey,
+    ) -> Result<()> {
+        let collected_client_data: CollectedClientData =
+            serde_json::from_slice(self.client_data_json.as_slice())?;
+        let challenge_bytes = Bytes::try_from(collected_client_data.challenge.as_str())
+            .map_err(|e| anyhow!("Failed to decode challenge bytes {:?}", e))?;
+
+        // Check that SHA3-256 of message is equal to challenge in client_data_json
+        let is_equal = verify_expected_challenge_from_message_matches_actual(
+            message,
+            challenge_bytes.as_slice(),
+        )?;
+
+        // Check if expected challenge and actual challenge match. If there's no match, throw error
+        if !is_equal {
+            return Err(anyhow!(
```

I think David is saying have verify_expected_challenge_from_message_matches_actual Err instead of returning bool and throwing Err based on the bool

---

Ah i see makes sense

---

Done